server/src/main/java/org/elasticsearch/gateway/PersistedClusterStateService.java

```diff
@@ -520,30 +521,62 @@ public void close() throws IOException {
         private final BigArrays bigArrays;
 
         boolean fullStateWritten = false;
+        private final AtomicBoolean closed = new AtomicBoolean();
 
         private Writer(List<MetaDataIndexWriter> metaDataIndexWriters, String nodeId, BigArrays bigArrays) {
             this.metaDataIndexWriters = metaDataIndexWriters;
             this.nodeId = nodeId;
             this.bigArrays = bigArrays;
         }
 
+        private void ensureOpen() {
+            if (closed.get()) {
+                throw new AlreadyClosedException("cluster state writer is closed already");
+            }
+        }
+
+        public boolean isOpen() {
+            return closed.get() == false;
+        }
+
+        private void closeIfAnyIndexWriterHasTragedyOrIsClosed() {
+            if (metaDataIndexWriters.stream().map(writer -> writer.indexWriter)
+                .anyMatch(iw -> iw.getTragicException() != null || iw.isOpen() == false)) {
```

Comment:
I'm surprised we need to check for a tragic exception ourselves here, since the `IndexWriter` checks this already. Can you clarify?

Reply:
I've looked at IndexWriter and concluded that there might be cases where `onTragicEvent` is called but where there is no corresponding call to `maybeCloseOnTragicEvent`, hence the extra safeguard here. @jpountz might know more why that's the case?

Reply:
I'm not very familiar with this part of the Lucene codebase, but comments and assertions suggest it might be due to the fact that closing could cause deadlocks if done during a flush. @s1monw might know more.

Reply:
Ok I see what you mean, let's leave it like this. This is trappier than I expected.